mount: replace mountinfo handling with moby/sys/mountinfo

thaJeztah opened the pull request:
Trying to reduce duplicated effort in maintaining a mountinfo parser, this patch replaces the local implementation with the implementation in github.com/moby/sys, which is actively maintained and contains various optimizations.


~Using a temporary vendor, pending https://github.com/moby/sys/pull/32 and https://github.com/moby/sys/pull/34~ merged

~With this change, the only consumer of `mount.Self()` is containerd/cri. `mount.PID()` looks to be unused altogether, so both could potentially be removed if containerd/cri is updated.~ removed

fuweid commented:
But it brings new dependencies into containerd 😂

estesp commented:
New dependencies, but fewer total lines of code? "+5,059 −8,426" ❓ 😂

thaJeztah commented:
> New dependencies, but fewer total lines of code? "+5,059 −8,426" ❓ 😂

Yeah, most of that is currently because of golang.org/x/sys and containerd/console. If we tag a new release of containerd/console, I can vendor it separately (nudge nudge 😉)

kolyshkin commented on mount/lookup_unix.go:
```diff
@@ -20,34 +20,21 @@ package mount
 	dir = filepath.Clean(dir)
 
-	mounts, err := Self()
+	m, err := mountinfo.GetMounts(mountinfo.SingleEntryFilter(dir))
```
Alas, this is wrong. Previously this function was capable of returning info for e.g. mount point `/mnt` given the argument of `/mnt/some/thing`. Currently it looks for `/mnt/some/thing`.

You need `mountinfo.ParentsFilter(dir)`, sort, and take the longest one -- same as in `getSourceMount` in moby/moby.

kolyshkin replied:
I also think we can implement this in mountinfo itself, as it's been used from a few places.

thaJeztah replied:
updated; PTAL

thaJeztah commented:
Moving this out of draft; removed the remaining uses in the CRI code now that it was integrated

@estesp @kolyshkin @cpuguy83 PTAL

kolyshkin commented on mount/mountinfo.go:
```diff
@@ -54,3 +56,27 @@ type Info struct {
+		Optional:   info.Optional,
+		FSType:     info.FSType,
+		Source:     info.Source,
+		VFSOptions: info.VFSOptions,
```
OK, now the fields are the same, so why don't we do something like

```go
type Info = mountinfo.Info
```

and drop the `fromMountinfo` entirely?

wdyt @thaJeztah?

kolyshkin replied:
Ahh, my bad, I see you're doing it in the next commit. Not sure if this intermediate step makes sense (iow I think this commit and the next one may need to be merged)

thaJeztah replied:
Ah, good one; I think it was separate because I had a temporary vendor in between, but yes, I can see it now doesn't make sense to implement (and remove) the conversion code in a single PR; I can squash them.

thaJeztah replied:
squashed the commits

kolyshkin commented on pkg/cri/server/helpers_linux.go:
```diff
@@ -161,7 +162,7 @@ func openLogFile(path string) (*os.File, error) {
 // the deepest mount first.
 func unmountRecursive(ctx context.Context, target string) error {
-	mounts, err := mount.Self()
+	mounts, err := mountinfo.GetMounts(nil)
```
Two things.

1. This might benefit from using PrefixFilter (and when the logic will be simplified -- there's no need to do filepath.Rel and strings.HasPrefix below).

2. Not really related to this, but there's a little known fact that umount(MNT_DETACH) is actually recursive in Linux, IOW this function can be replaced with `unix.Umount(target, unix.MNT_DETACH)` (or `mount.UnmountAll(target, unix.MNT_DETACH)`) -- provided that `target` itself is a mount point. Related: https://github.com/moby/sys/pull/26

3. Definitely not related to this, but I took a look at `pkg/mount.UnmountAll` and :open_mouth: it retries umount without a delay on any error rather than EINVAL forever. This means, for example, if someone denies umount syscall using seccomp, we'll end up calling umount in a busy loop indefinitely.

thaJeztah replied:
@kolyshkin would that mean we can replace this whole function with `moby/mount.RecursiveUnmount()`, or would we still need some of the logic below?

thaJeztah replied:
> it retries umount without a delay on any error rather than EINVAL

I see it returns on errors (only returns `nil` if it was an `EINVAL`), or is that not the part you were looking at? https://github.com/containerd/containerd/blob/45d8a7e4a2d1e69dad4e1d879a7a7210a6d7ec71/mount/mount_linux.go#L179-L182

kolyshkin replied:
> I see it returns on errors (only returns nil if it was an EINVAL), or is that not the part you were looking at?

You're right, I misread the code in there :) so item 3 above can be crossed out.

Re. item 2, I finally understood the logic of containerd's `UnmountAll()` (it works around overmounts, i.e. when the same mountpoint being mounted to multiple times) and I think we should do something like this in moby/sys/mount, too. In the meantime, let's leave this as is, IOW ACK.

kolyshkin commented:
@thaJeztah thanks for working on this! LGTM overall, maybe makes sense to squash the first two (or three?) commits.

thaJeztah commented:
@kolyshkin I added the prefix-filter, and removed the local prefix checks; kept it as a separate commit for easier reviewing, but I can squash.

> IOW this function can be replaced with unix.Umount(target, unix.MNT_DETACH) (or mount.UnmountAll(target, unix.MNT_DETACH)) 

Do you think we should replace it here? (mostly wondering if the current logic was addressing specific (corner) cases)

thaJeztah commented:
Rebased to get a fresh run of CI

@AkihiroSuda @estesp @kolyshkin @cpuguy83 PTAL

kolyshkin commented on mount/mountinfo.go:
```diff
@@ -16,41 +16,16 @@
-	VFSOptions string
+type Info = mountinfo.Info
+
+func fromMountinfoSlice(infos []*mountinfo.Info) []Info {
```
As the types (`mountinfo.Info` and `Info`) are now identical, it seems we no longer need this function.

kolyshkin replied:
Or is it because moby/sys.Mountinfo returns a slice of pointers? In this case, I think, it will be better to change the functions here to also return a slice of pointers. It won't result in any changes to the callers (as far as I can see).

kolyshkin replied:
Nevermind -- I see you do this in the next commit :)

kolyshkin commented on mount/mountinfo_bsd.go:
```diff
@@ -18,46 +18,20 @@
-	}
-	return out, nil
+	m, err := mountinfo.GetMounts(nil)
+	return fromMountinfoSlice(m), err
```
i.e.
```suggestion
	return mountinfo.GetMounts(nil)
```

kolyshkin commented on mount/temp_unix.go:
```diff
@@ -40,23 +41,20 @@ func SetTempMountLocation(root string) error {
 // CleanupTempMounts all temp mounts and remove the directories
 func CleanupTempMounts(flags int) (warnings []error, err error) {
-	mounts, err := Self()
+	mounts, err := mountinfo.GetMounts(mountinfo.PrefixFilter(tempMountLocation))
```
Alas we can't have a "fast path" I talked about earlier in here, since `tempMountLocation` is not a mount point per se.

In case containerd itself is doing all those mounts, those can be saved in a list, and we should go over that list in here (instead of reading mountinfo) -- that would be a worthwhile optimization.

In case a user does those mounts, this is unfortunate and there's nothing we can do but read mountinfo.

In any case, this can be addressed later.

kolyshkin reviewed (approved):
LGTM

dmcgowan reviewed (approved):
LGTM

estesp reviewed (approved):
LGTM
